### PR TITLE
Escape all user input that is part of an NQL query

### PR DESCRIPTION
**Pull request by kgodlewski**
This PR makes all queries that use NQL under the hood be consistent with how UI works. **This is a breaking change.**

The current behaviour forces users to manually escape backslash and quotes when providing arguments to parameters that expect a regex, example:

```python
# Match all experiments with names "experiment-0", "experiment-1", etc
df = project.fetch_experiments_df(names_regex=r"experiment-\\d")
```

We want the behaviour to be consistent with the UI, eg. not require any escaping:

```python
# Yields the same result as the old version above
df = project.fetch_experiments_df(names_regex=r"experiment-\d")
```

Note that this does not change how filtering attribute names works -- it's already correct, since the change was only required for arguments that make it into NQL.

```python
# No change required
df = project.fetch_experiments_df(columns_regex=r"metrics/foo\d")
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -148,6 +148,13 @@ print(run["metrics/accuracy"].fetch_values())
 
 ## API reference
 
+### Supported regular expressions
+
+Neptune uses the [RE2](https://github.com/google/re2) regular expression library. Please refer to
+the official [syntax guide](https://github.com/google/re2/wiki/syntax) for more information on
+the supported regex features and limitations.
```
Comment (normandy7) [inline README.md:155]:
```suggestion
Neptune uses the [RE2](https://github.com/google/re2) regular expression library. For supported regex features and limitations, see the official [syntax guide](https://github.com/google/re2/wiki/syntax).
```

File: README.md [hunk 2]
```diff
@@ -266,7 +273,7 @@ __Parameters:__
...
-| `query`                    | `str`, optional                               | `None`              | NQL query string. Example: `"(accuracy: float > 0.88) AND (loss: float < 0.2)"`. The query is applied on top of other criteria like, `custom_ids`, `tags` etc, using the logical AND operator. For syntax, see [Neptune Query Language](https://docs.neptune.ai/usage/nql/) in Neptune docs.                                                       |
+| `query`                    | `str`, optional                               | `None`              | NQL query string. Example: `"(accuracy: float > 0.88) AND (loss: float < 0.2)"`. The query is applied on top of other criteria like, `custom_ids`, `tags` etc, using the logical AND operator. Please see the examples below. For syntax, see [Neptune Query Language](https://docs.neptune.ai/usage/nql/) in Neptune docs.                                                       |
```
Comment (normandy7) [inline README.md:276]:
```suggestion
| `query`                    | `str`, optional                               | `None`              | NQL query string. Example: `"(accuracy: float > 0.88) AND (loss: float < 0.2)"`. The query is applied on top of other criteria like, `custom_ids`, `tags` etc, using the logical AND operator. See examples below. For syntax, see [Neptune Query Language](https://docs.neptune.ai/usage/nql/) in the Neptune docs.                                                       |
```

File: README.md [hunk 3]
```diff
@@ -314,7 +321,7 @@ specific_runs_df = my_project.fetch_runs_df(
 )
 ```
 
-Fetch runs with a complex query:
+Fetch runs with a complex query using NQL.
```
Comment (normandy7) [inline README.md:324]:
```suggestion
Fetch runs with a complex query using NQL:
```

File: README.md [hunk 4]
```diff
@@ -348,7 +355,7 @@ __Parameters__:
...
-| `query`                    | `str`, optional                               | `None`              | NQL query string. Example: `"(accuracy: float > 0.88) AND (loss: float < 0.2)"`. The query is applied on top of other criteria like, `names_regex`, `tags` etc, using the logical AND operator. For syntax, see [Neptune Query Language](https://docs.neptune.ai/usage/nql/) in Neptune docs.                                                      |                                                          |
+| `query`                    | `str`, optional                               | `None`              | NQL query string. Example: `"(accuracy: float > 0.88) AND (loss: float < 0.2)"`. The query is applied on top of other criteria like, `custom_ids`, `tags` etc, using the logical AND operator. Please see the examples below. For syntax, see [Neptune Query Language](https://docs.neptune.ai/usage/nql/) in Neptune docs.                                                       |
```
Comment (normandy7) [inline README.md:358]:
```suggestion
| `query`                    | `str`, optional                               | `None`              | NQL query string. Example: `"(accuracy: float > 0.88) AND (loss: float < 0.2)"`. The query is applied on top of other criteria like, `custom_ids`, `tags` etc, using the logical AND operator. See examples below. For syntax, see [Neptune Query Language](https://docs.neptune.ai/usage/nql/) in the Neptune docs.                                                       |
```

File: README.md [hunk 5]
```diff
@@ -388,11 +395,11 @@ specific_experiments_df = my_project.fetch_experiments_df(
 )
 ```
 
-Fetch experiments with a complex query:
+Fetch experiments with a complex query using NQL. Note that we're using a raw Python string, and the `\` character is escaped to be properly interpreted.
 
 ```python
```
Comment (normandy7) [inline README.md:400]:
```suggestion
Fetch experiments with a complex query using NQL.

> **Note:** Because we're using a raw Python string, the `\` character must be escaped.

```python
```

File: README.md [hunk 6]
```diff
@@ -395,7 +393,16 @@ specific_experiments_df = my_project.fetch_experiments_df(
...
+Fetch experiments with a complex query using NQL. Note that the `\` character needs to be escaped in the resulting NQL query
+for the backend to interpret the argument correctly.
+
+```python
+experiments_df = my_project.fetch_experiments_df(
+    query='(`learning_rate`:float < 0.01) AND (`sys/name`:string MATCHES "experiment-\\\\d+")'
+)
+```
+
+Using raw Python strings is a less cluttered alternative:
```
Comment (normandy7) [inline README.md:405]:
```suggestion
Use the Neptune Query Language to fetch experiments with a complex query. Note that for regular strings, the `\` character needs to be escaped:

```python
experiments_df = my_project.fetch_experiments_df(
    query='(`learning_rate`:float < 0.01) AND (`sys/name`:string MATCHES "experiment-\\\\d+")'
)
```

As a less cluttered alternative, you can pass a raw Python string to the `query` argument:

```

**Review by PatrykGala — APPROVED** [review]
